### ISSUE TITLE
Add monitoring feature

### ISSUE THREAD

**Issue by vladimir-bukhtoyarov** [vladimir-bukhtoyarov/bucket4j#12]
Following points should be exposed:
- Token consumption.
- Token rejection.
- Amount of time spent in waiting to bucket refill.
- Count of InterruptedExceptions happens during waiting.

Design considerations:
- Monitoring must be pluggable, most likely statistic should be implemented as listener with methods onConsumed, onRejected, onWait, onInterrupt, it should be responsibility of client to decide how to this data should be accumulated, aggregated and exposed.
- Reference implementation on top of [Dropwizard-Metrics](http://metrics.dropwizard.io/3.1.0/manual/core/) should be provided.

Also it is need to support requqirements from #7

**Comment by vladimir-bukhtoyarov** [comment]
Looks like it would be better if end user will implement monitoring by itself, with preferred monitoring library.

**Comment by bfalese-navent** [comment]
Hello,
I'm just starting to use bucket4j and i would like to record usage metrics. 

Are this listening points (onConsumed, onRejected, onWait, onInterrupt) mentioned in this thread already available or where a proposed solution?

If this extension already exist could you point me in the right direction?
Thank you

**Comment by vladimir-bukhtoyarov** [comment]
Hello @bfalese-navent

The current issue is closed because it is impossible to implement monitoring which will be suitable for all cases.

But I will try to point the solution which suitable for your particular case if you will explain following points:
- Your usage scenario: do you use distributed bucket or just in-memory? Do you use many buckets or just single one?
- The monitoring library which you use to expose/export metrics
- Monitoring database and visualization tool.

**Comment by bfalese-navent** [comment]
Hello @vladimir-bukhtoyarov 

The current usage scenario is:
- In-memory with many buckets.
- Micrometer.io, using dropwizard metrics as the underlying reporter, through jmx.
- Zabbix as metrics collector and visualization tool.
 
Thank for your time.

**Comment by vladimir-bukhtoyarov** [comment]
I will provide solution tommorrow

**Comment by vladimir-bukhtoyarov** [comment]
I created a concept of monitoring. Four types of events are measuring:
```java
public interface BucketListener {

    void onConsumed(long tokens);

    void onRejected(long tokens);

    void onParked(long nanos);

    void onInterrupted();

}
```

To consume the events it is necessary to decorate the original bucket:
```java
bucket = ListeneableBucket.decorate(bucket, listener);
```

If you want to have aggregated statistics for all buckets then create single listener per application and reuse this listener to decorate all buckets. If you want to measure statistics independently per each bucket then use listener per bucket model.

It is simple implementation of listener:
```java
import java.util.concurrent.atomic.AtomicLong;

public class SimpleBucketListener implements BucketListener {

    private AtomicLong consumed = new AtomicLong();
    private AtomicLong rejected = new AtomicLong();
    private AtomicLong parkedNanos = new AtomicLong();
    private AtomicLong interrupted = new AtomicLong();

    @Override
    public void onConsumed(long tokens) {
        consumed.addAndGet(tokens);
    }

    @Override
    public void onRejected(long tokens) {
        rejected.addAndGet(tokens);
    }

    @Override
    public void onParked(long nanos) {
        parkedNanos.addAndGet(nanos);
    }

    @Override
    public void onInterrupted() {
        interrupted.incrementAndGet();
    }

    public long getConsumed() {
        return consumed.get();
    }

    public long getRejected() {
        return rejected.get();
    }

    public long getParkedNanos() {
        return parkedNanos.get();
    }

    public long getInterrupted() {
        return interrupted.get();
    }

}
```

It is example of exposing statistics via Dropwizard Metrics(for Micrometer it should be quite similar)
```java
public static void registerBucketStat(SimpleBucketListener stat, MetricRegistry registry, String name) {
  registry.register(name + ".consumed", (Gauge<Long>) -> stat::getConsumed);
  registry.register(name + ".rejected", (Gauge<Long>) -> stat::getRejected);
  registry.register(name + ".parkedNanos", (Gauge<Long>) -> stat::getParkedNanos);
  registry.register(name + ".interrupted", (Gauge<Long>) -> stat::getInterrupted);
}
```

This is code of ListeneableBucket
```java
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ScheduledExecutorService;

public class ListeneableBucket implements Bucket {

    private final BucketListener listener;
    private final Bucket target;

    public static Bucket decorate(Bucket target, BucketListener listener) {
        return new ListeneableBucket(target, listener);
    }

    private ListeneableBucket(Bucket target, BucketListener listener) {
        this.listener = Objects.requireNonNull(listener);
        this.target = Objects.requireNonNull(target);
    }

    @Override
    public boolean isAsyncModeSupported() {
        return target.isAsyncModeSupported();
    }

    @Override
    public boolean tryConsume(long numTokens) {
        if (tryConsume(numTokens)) {
            listener.onConsumed(numTokens);
            return true;
        } else {
            listener.onRejected(numTokens);
            return false;
        }
    }

    @Override
    public ConsumptionProbe tryConsumeAndReturnRemaining(long numTokens) {
        ConsumptionProbe probe = target.tryConsumeAndReturnRemaining(numTokens);
        if (probe.isConsumed()) {
            listener.onConsumed(numTokens);
        } else {
            listener.onRejected(numTokens);
        }
        return probe;
    }

    @Override
    public long tryConsumeAsMuchAsPossible() {
        long consumed = target.tryConsumeAsMuchAsPossible();
        if (consumed > 0) {
            listener.onConsumed(consumed);
        }
        return consumed;
    }

    @Override
    public long tryConsumeAsMuchAsPossible(long limit) {
        long consumed = target.tryConsumeAsMuchAsPossible(limit);
        if (consumed > 0) {
            listener.onConsumed(consumed);
        }
        return consumed;
    }

    @Override
    public boolean tryConsume(long numTokens, long maxWaitTimeNanos, BlockingStrategy blockingStrategy) throws InterruptedException {
        boolean consumed;
        try {
            consumed = target.tryConsume(numTokens, maxWaitTimeNanos, new BlockingStrategy() {
                @Override
                public void park(long nanosToPark) throws InterruptedException {
                    blockingStrategy.park(nanosToPark);
                    listener.onParked(nanosToPark);
                }

                @Override
                public void parkUninterruptibly(long nanosToPark) {
                    // TODO consider to split BlockingStrategy by two independent interfaces
                    throw new IllegalStateException("should not be called");
                }
            });
        } catch (InterruptedException e) {
            listener.onInterrupted();
            listener.onRejected(numTokens);
            throw e;
        }
        if (consumed) {
            listener.onConsumed(numTokens);
        } else {
            listener.onRejected(numTokens);
        }
        return consumed;
    }

    @Override
    public boolean tryConsumeUninterruptibly(long numTokens, long maxWaitTimeNanos, BlockingStrategy blockingStrategy) {
        boolean consumed = target.tryConsumeUninterruptibly(numTokens, maxWaitTimeNanos, new BlockingStrategy() {
            @Override
            public void park(long nanosToPark) throws InterruptedException {
                // TODO consider to split BlockingStrategy by two independent interfaces
                throw new IllegalStateException("should not be called");
            }

            @Override
            public void parkUninterruptibly(long nanosToPark) {
                blockingStrategy.parkUninterruptibly(nanosToPark);
                listener.onParked(nanosToPark);

            }
        });

        if (consumed) {
            listener.onConsumed(numTokens);
        } else {
            listener.onRejected(numTokens);
        }
        return consumed;
    }

    @Override
    public void addTokens(long tokensToAdd) {
        target.addTokens(tokensToAdd);
    }

    @Override
    public long getAvailableTokens() {
        return target.getAvailableTokens();
    }

    @Override
    public void replaceConfiguration(BucketConfiguration newConfiguration) {
        target.replaceConfiguration(newConfiguration);
    }

    @Override
    public BucketState createSnapshot() {
        return target.createSnapshot();
    }

    private final AsyncBucket asyncView = new AsyncBucket() {
        @Override
        public CompletableFuture<Boolean> tryConsume(long numTokens) {
            return target.asAsync().tryConsume(numTokens).thenApply(consumed -> {
                if (consumed) {
                    listener.onConsumed(numTokens);
                } else {
                    listener.onRejected(numTokens);
                }
                return consumed;
            });
        }

        @Override
        public CompletableFuture<ConsumptionProbe> tryConsumeAndReturnRemaining(long numTokens) {
            return target.asAsync().tryConsumeAndReturnRemaining(numTokens).thenApply(probe -> {
                if (probe.isConsumed()) {
                    listener.onConsumed(numTokens);
                } else {
                    listener.onRejected(numTokens);
                }
                return probe;
            });
        }

        @Override
        public CompletableFuture<Long> tryConsumeAsMuchAsPossible() {
            return target.asAsync().tryConsumeAsMuchAsPossible().thenApply(consumedTokens -> {
                if (consumedTokens > 0) {
                    listener.onConsumed(consumedTokens);
                }
                return consumedTokens;
            });
        }

        @Override
        public CompletableFuture<Long> tryConsumeAsMuchAsPossible(long limit) {
            return target.asAsync().tryConsumeAsMuchAsPossible(limit).thenApply(consumedTokens -> {
                if (consumedTokens > 0) {
                    listener.onConsumed(consumedTokens);
                }
                return consumedTokens;
            });
        }

        @Override
        public CompletableFuture<Boolean> tryConsume(long numTokens, long maxWaitNanos, ScheduledExecutorService scheduler) {
            return target.asAsync().tryConsume(numTokens, maxWaitNanos, scheduler).thenApply(consumed -> {
                if (consumed) {
                    listener.onConsumed(numTokens);
                } else {
                    listener.onRejected(numTokens);
                }
                return consumed;
            });
        }

        @Override
        public CompletableFuture<Void> addTokens(long tokensToAdd) {
            return target.asAsync().addTokens(tokensToAdd);
        }

        @Override
        public CompletableFuture<Void> replaceConfiguration(BucketConfiguration newConfiguration) {
            return target.asAsync().replaceConfiguration(newConfiguration);
        }
    };

    @Override
    public AsyncBucket asAsync() {
        if (!isAsyncModeSupported()) {
            throw new UnsupportedOperationException();
        }
        return asyncView;
    }

}
```
It was not tested, but I hope that it works properly.

@bfalese-navent feel free to provide feedback or push enhancements into ```monitoring``` branch. Eventually, this branch will be released, but currently I have no time to do so, because of I am busy with next generation of my another library [Rolling-Metrics](https://github.com/vladimir-bukhtoyarov/rolling-metrics).

**Comment by bfalese-navent** [comment]
Hello @vladimir-bukhtoyarov 

Wow!, Thank you very much for your time. 

The solution and explanation looks perfect, i will be using it as is.

Excellent library Rolling-Metrics, i'm starting to use smoothly decaing rolling window counter.

**Comment by vladimir-bukhtoyarov** [comment]
Done. Documentation can be found there https://github.com/vladimir-bukhtoyarov/bucket4j/blob/4.0/doc-pages/listener.md

will be released soon within ```4.0``` train.

**Comment by vladimir-bukhtoyarov** [comment]
released https://github.com/vladimir-bukhtoyarov/bucket4j/releases/tag/4.0.0

**Comment by MarcGiffing** [comment]
It would be nice if you can provide additonal informations on consuming tokens. Such as the user, the country, the IP Address or the URL so that these informations can be used in the BucketListener for the monitoring.

For Example:
```java
 bucket.tryConsumeAndReturnRemaining(1, "country", "spain")

public interface BucketListener {
    void onConsumed(long tokens, String[] tags);
    void onRejected(long tokens, String[] tags);
}
```

https://micrometer.io/docs/concepts#_tag_naming

Is something like that possible?

**Comment by vladimir-bukhtoyarov** [comment]
Hello Marc,

Extending only one method  ```tryConsumeAndReturnRemaining``` would be inconsistent. There are many similar methods inside Bucket, AsyncBucket, AsyncScheduledBucket and BlockingBucket interfaces, there are many efforts to extend all these interfaces.

I recommend you to cover the tagging problem in different ways:
* If you do not use in-memory bucket then just call ```toListenable``` before each request to bucket:
```java
Map<String,String> tags = ...;
BucketListener listener = new TagsAwareListener(tags);
bucket.toListenable(listener).tryConsume...
```
The fact that ```toListenable``` creates new instance of bucket is not a problem, because BucketProxy does not hold the state of bucket, all things will work correctly independently from how many copies of proxies you created, because the actual state stored somewhere outside.
* When in-memory buckets is used the things became more complicated, because in opposite to distributed bucket for in-memory the ```toListenable``` creates the new independent copy of bucket, so sulution above will not work. Lets find another way, in case of one bucket per one user model(tags are specific for bucket only), just add all tags related with user directly to listener, in other words use different listeners for different buckets.
* If many users(with different associated tags) share the same bucket(tags are specific for particular request) then ThreadLocal can help - just remember tags in threadlocal and then read from threadlocal inside listener, it is obviously then one listener instance would be enough. Unfortunately threadlocals will not work correctly with  async API, if this is not applicable then Bucket4j is not able to help you, but it is still possible to collect the stat by yourself, of coarse you will need write the significant amount of code like there https://github.com/vladimir-bukhtoyarov/bucket4j/issues/12#issuecomment-364643775

**The best solution from my point of view**: will be eliminating difference between in-memory and distributed buckets, similar to distributed the implementation of ```toListanable`` for in-memory must create new bucket which shares the state with previous bucket. I have created the issue for this #83. But currently I have no time to implement this stuff, because I heavy concentrated on implementing support for Redis. Anyway #83 will be release in scope of 5.0(in same time with Redis). I have no plans to fix #83 prior ```5.0``` because as mentioneed above you are able to use work-around https://github.com/vladimir-bukhtoyarov/bucket4j/issues/12#issuecomment-364643775

**Comment by vladimir-bukhtoyarov** [comment]
Hi Marc,

I have release version ```4.1.1``` which fixes #83(it was easy exercise), so the best way to cover your needs would be decorate bucket with new Listener on each request via ```toListenable```.

**Comment by MarcGiffing** [comment]
Thank you very much!
